### PR TITLE
Test for WindowFrame response of WindowSpec of Apache Spark for Preceding and Following values

### PR DESCRIPTION
**Pull request by udurbha**
Screenshot 2022-11-21 5 59 47 PM
The windowFrame property of WindowSpec in Apache Spark returns a response 
eg: " 1 Following and 2 Following".   In the above screen shot of CDAP Window Aggregation Analytics, we have options to choose 
1.Unbounded Preceding and VALUE Following 
2.VALUE Preceding and  Unbounded Following 
3.VALUE Preceding and VALUE Following
4.Unbounded Preceding and Unbounded Following .
But Not
1.VALUE Preceding and VALUE Preceding
2.VALUE Following and VALUE Following
Since WindowFrame property is private in Window Spec ,to assert the response used reflect method.
 **Positive** number in **preceding** field input  gives the  response **NUMBER** **following** and **Negative** number in **following** field input gives response **NUMBER** **preceding** This test case is to make sure that we are aware of the current scenario where a positive preceding value is taken as a following and negative following value is taken as preceding.Also removed the dependencies in pom.xml which are not necessary

### CODE REVIEW

File: src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java [hunk 1]
```diff
@@ -190,4 +194,20 @@ public void testInvalidParameter() {
                                                                           config, schema, schema);
     }
   }
+
+  @Test
+  public void testWindowSpec() throws NoSuchFieldException, IllegalAccessException {
+    WindowSpec spec = Window.partitionBy(WindowsAggregationUtil.getPartitionsColumns(config.getPartitionFields()))
+                            .orderBy(WindowsAggregationUtil.getPartitionOrderColumns(config.getPartitionOrder()));
+    spec = spec.rangeBetween(config.getFrameDefinitionPrecedingBound(),
```
Comment (DJSagarAhire) [inline src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java]: I *think* I found the issue.

As per the [javadoc](https://spark.apache.org/docs/2.3.0/api/java/org/apache/spark/sql/expressions/WindowSpec.html#rangeBetween-long-long-), the argument to the `rangeBetween()` function is a number of rows offset from the current row, where 0 is the current row, a positive number represents a row following the current row and a negative number represents a row preceding the current row. Therefore, if we want to have a window with rows preceding the current row, the row number should be **negative**.

Spark source code where this is handled is present in [windowExpressions.scala](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala#L271).

TL;DR: Use a negative number for `PRECEDING`.

Comment (udurbha) [reply src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java]: Yes .For preceding negative number is taken.IF it is positive number in preceding input it gives the current response.This test case is to make sure that we are aware of the current scenario where a **positive** preceding value is taken as a **following**.We can get confirmation from the user that preceding will be converted as following if its a positive value

File: src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java [hunk 2]
```diff
@@ -139,14 +144,14 @@ public void setUp() {
     Mockito.when(sparkExecutionPluginContext.getFailureCollector()).thenReturn(failureCollector);
     sparkContext = Mockito.mock(SparkContext.class);
     Mockito.when(javaSparkContext.sc()).thenReturn(sparkContext);
-    List<Field> schemaFields = new ArrayList<>();
+    List<Schema.Field> schemaFields = new ArrayList<>();
```
Comment (fernst) [inline src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java]: Why this change?

File: src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java [hunk 3]
```diff
@@ -139,14 +144,14 @@ public void setUp() {
...
     Field field = Field.of("field", Schema.of(Schema.Type.INT));
     schemaFields.add(field);
     schema = recordOf("test", schemaFields);
 
     Mockito.when(data.withColumn(Mockito.anyString(), Mockito.any(Column.class))).thenReturn(data);
     structSchema = Mockito.mock(StructType.class);
-    Mockito.when(config.getWindowFrameType()).thenReturn(WindowAggregationConfig.WindowFrameType.ROW);
+    Mockito.when(config.getWindowFrameType()).thenReturn(WindowAggregationConfig.WindowFrameType.RANGE);
```
Comment (fernst) [inline src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java]: We should be testing all possible configurations here

File: src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java [hunk 4]
```diff
@@ -160,6 +165,7 @@ public void setUp() {
 
   @Test
   public void testTransform() {
+    int flag=0;
```
Comment (fernst) [inline src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java]: This isn't being used anywhere.

File: src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java (hunk 2 above)
Comment (udurbha) [reply src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java]: Initially there was another Field class, due to which specifically it was mentioned.Removed it

File: src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java (hunk 3 above)
Comment (udurbha) [reply src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java]: Done

File: src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java (hunk 4 above)
Comment (udurbha) [reply src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java]: Removed

File: src/main/java/io/cdap/plugin/WindowsAggregationUtil.java [hunk 5]
```diff
@@ -197,8 +197,9 @@ private static Schema schemaForInputField(Schema inputSchema, String inputFieldN
     return schemaField.getSchema();
   }
 
-  private static Column[] getPartitionsColumns(List<String> partitionsFields) {
-    return partitionsFields.stream().map(Column::new).toArray(Column[]::new);
+  public static Column[] getPartitionsColumns(List<String> partitionsFields) {
+    Column[] partitionColumns = partitionsFields.stream().map(Column::new).toArray(Column[]::new);
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/WindowsAggregationUtil.java]: You can revert this change.

File: pom.xml [hunk 6]
```diff
@@ -182,6 +182,11 @@
       <version>4.8.1</version>
       <scope>test</scope>
     </dependency>
+      <dependency>
```
Comment (fernst) [inline pom.xml]: Do you need this dependency? If absolutely needed, please add a `scope` of `test` for this dependency. It shouldn't be used when compiling

**Review by fernst — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 6 above)
Comment (fernst) [reply pom.xml]: Also use Scala 2.11 version instead of scala 2.10

File: src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java (hunk 1 above)
Comment (udurbha) [reply src/test/java/io/cdap/plugin/WindowsAggregationUtilTest.java]: Also for the field **following** if a **negative** value is entered it takes as **Preceding** of that number

File: src/main/java/io/cdap/plugin/WindowsAggregationUtil.java (hunk 5 above)
Comment (udurbha) [reply src/main/java/io/cdap/plugin/WindowsAggregationUtil.java]: Done

File: pom.xml (hunk 6 above)
Comment (udurbha) [reply pom.xml]: Done

Comment (udurbha) [reply pom.xml]: > Do you need this dependency? If absolutely needed, please add a `scope` of `test` for this dependency. It shouldn't be used when compiling

Scope cannot be test as this dependency is used by org.apache.spark.sql.* in WindowAggregationUtil.java

Comment (fernst) [reply pom.xml]: Then how did this code compile before? You didn't make any changes to WindowsAggregatonUtil in this PR

File: src/main/java/io/cdap/plugin/WindowsAggregationUtil.java [hunk 7]
```diff
@@ -197,7 +197,7 @@ private static Schema schemaForInputField(Schema inputSchema, String inputFieldN
     return schemaField.getSchema();
   }
 
-  private static Column[] getPartitionsColumns(List<String> partitionsFields) {
+  public static Column[] getPartitionsColumns(List<String> partitionsFields) {
```
Comment (DJSagarAhire) [inline src/main/java/io/cdap/plugin/WindowsAggregationUtil.java:202]: I believe this is made public only for testing purposes. Please add a `@VisibleForTesting` annotation in that case.

Comment (udurbha) [reply src/main/java/io/cdap/plugin/WindowsAggregationUtil.java:202]: Done

**Review by DJSagarAhire — APPROVED** [review]
LGTM
